Mathlib/SetTheory/Ordinal/Club.lean

```diff
@@ -0,0 +1,263 @@
+  rcases h.2 x (hx ▸ lt_succ x) with ⟨p, hp⟩
+  exact (hx ▸ succ_le_iff.mpr hp.2).not_lt (h.1 hp.1)
+
+theorem IsUnbounded.inter_Ioo_nonempty {o : Ordinal} {U} (hU : IsUnbounded U o) {p : Ordinal}
+    (hp : p < o) : (U ∩ Ioo p o).Nonempty :=
+  (hU.2 p hp).casesOn (fun x ⟨xmemU, pltx⟩ ↦ ⟨x, xmemU, ⟨pltx, hU.1 xmemU⟩⟩)
+
+
+section ClubIntersection
+
+variable {o : Ordinal.{u}} {S : Set (Set Ordinal)}
+variable {ι : Type u} {f : ι → Set Ordinal}
+
+theorem IsClosed.sInter (hS : S.Nonempty) (h : ∀ C ∈ S, IsClosed C o) : IsClosed (⋂₀ S) o where
+  left s hs := by
+    rcases hS with ⟨C, CmemS⟩
+    exact (h C CmemS).1 ((sInter_subset_of_mem CmemS) hs)
+  right p plto pAcc := by
+    rw [mem_sInter]
+    intro C CmemS
+    exact (h C CmemS).2 p plto (pAcc.subset (sInter_subset_of_mem CmemS))
+
+theorem IsClosed.iInter [Nonempty ι] (h : ∀ i, IsClosed (f i) o) :
+    IsClosed (⋂ i, f i) o := by
+  have := IsClosed.sInter (range_nonempty f) (fun _ ⟨x, hx⟩ ↦ hx ▸ h x)
+  rwa [sInter_range] at this
+
+/-- Given less than `o.cof` unbounded sets in `o` and some `q < o`, there is a `q < p < o`
+  such that `Ioo q p` contains an element of every unbounded set. -/
+theorem exists_above_of_lt_cof {p : Ordinal} (hp : p < o) (hSemp : Nonempty S)
+    (hSunb : ∀ U ∈ S, IsUnbounded U o) (hScard : #S < Cardinal.lift.{u + 1, u} o.cof) :
+    ∃ q, q < o ∧ p < q ∧ ∀ U ∈ S, (U ∩ Ioo p q).Nonempty := by
+  rw [lift_cof] at hScard
+  have oLim : IsLimit o := hSemp.casesOn fun ⟨T, hT⟩ ↦
+    (hSunb T hT).isLimit (pos_of_gt hp).ne.symm
+  let f : ↑S → Ordinal := fun U ↦ lift.{u + 1, u} (sInf (U ∩ (Ioo p o)))
+  have infMem : ∀ U : S, sInf (↑U ∩ Ioo p o) ∈ ↑U ∩ Ioo p o := fun U ↦
+    csInf_mem ((hSunb U.1 U.2).inter_Ioo_nonempty hp : (↑U ∩ Ioo p o).Nonempty)
+  have flto : ∀ U : S, f U < lift.{u + 1, u} o := fun U ↦ by
+    simp_all only [mem_inter_iff, mem_Ioo, lift_lt, f]
+  set q := (Ordinal.sup.{u + 1, u} f) + 1 with qdef
+  have qlto : q < lift.{u + 1, u} o :=
+    ((lift_isLimit.{u + 1, u} o).mpr oLim).2 (sup.{u + 1, u} f) (sup_lt_ord hScard flto)
+  rcases lift_down qlto.le with ⟨q', hq'⟩
+  use q'
+  have fltq : ∀ U, f U < q := fun U ↦
+    lt_of_le_of_lt (le_sup.{u + 1, u} f U) (qdef ▸ lt_add_one (sup f))
+  constructor <;> try constructor
+  · exact lift_lt.mp (hq' ▸ qlto)
+  · rcases hSemp with ⟨U, hU⟩
+    have pltf : lift.{u + 1, u} p < f ⟨U, hU⟩ :=
+      lift_lt.mpr (mem_of_mem_inter_right (infMem ⟨U, hU⟩)).1
+    have := lt_of_lt_of_le pltf (fltq ⟨U, hU⟩).le
+    rwa [← hq', lift_lt] at this
+  intro U hU
+  specialize infMem ⟨U, hU⟩
+  specialize fltq ⟨U, hU⟩
+  have : f ⟨U, hU⟩ ∈ Ioo (lift.{u + 1, u} p) q := ⟨lift_lt.mpr infMem.2.1, fltq⟩
+  rw [← hq'] at fltq
+  rcases lift_down fltq.le with ⟨fUdown, fUlift⟩
+  use fUdown
+  constructor
+  · simp_all only [lift_inj, mem_inter_iff, f]
+  · constructor
+    exact lift_lt.mp <| fUlift ▸ (this.1)
+    exact lift_lt.mp <| hq' ▸ (fUlift ▸ this).2
+
+theorem strictMono_of_succ_lt_omega (f : Π p < ω, Iio o)
```

Comment:
I changed the proof to use `relIso_nat_omega`, but I do want to keep this one with `Π i < ω, α` because there are (and will be more) theorems about such sequences with length `δ` for a limit ordinal `δ`.